provider/source/src/time_zones/convert.rs

```diff
@@ -51,38 +51,50 @@ impl DataProvider<TimeZoneEssentialsV1Marker> for SourceDataProvider {
             .dates
             .time_zone_names;
 
-        fn parse_hour_format(hour_format: &str) -> (Cow<'static, str>, Cow<'static, str>) {
-            // e.g. "+HH:mm;-hh:mm" -> ("+H:m", "-h:m")
-            let index = hour_format.rfind(';').unwrap();
-            let positive = hour_format[0..index]
-                .replace("H", "h")
-                .replace("hh", "H")
-                .replace("mm", "m");
-            let negative = hour_format[index + 1..]
-                .replace("H", "h")
-                .replace("hh", "H")
-                .replace("mm", "m");
-            (Cow::Owned(positive), Cow::Owned(negative))
-        }
+        // e.g. "+HH:mm;-hh:mm"
+        let (offset_pattern_positive, offset_pattern_negative) =
+            time_zone_names.hour_format.split_once(';').unwrap();
+
+        // The single character before H/h
+        let offset_positive_sign = offset_pattern_positive
+            .split_once(['h', 'H'])
+            .unwrap()
+            .0
+            .chars()
+            .next()
+            .unwrap_or('+');
+        let offset_negative_sign = offset_pattern_negative
+            .split_once(['h', 'H'])
+            .unwrap()
+            .0
+            .chars()
+            .next()
+            .unwrap_or('-');
+        // The single character before `mm`
+        let offset_separator = offset_pattern_positive
+            .split_once('m')
+            .unwrap()
+            .0
+            .chars()
+            .next_back()
+            .unwrap_or(':');
 
```

Comment:
Nit: I'd prefer this pattern parsing to be more strict. If there's data that violates one of our invariants, we should know about it.

Reply:
#5680